### PR TITLE
WIP: Yajl: Removing empty parent keys

### PR DESCRIPTION
**Pull request by PhilippGackstatter**
Getting started with this pull request for #2132.
Also I don't know how to add labels like work in progress or if I just don't have the rights to do so.

## Basics

Check relevant points but **please do not remove entries**.
Do not describe the purpose of this PR in the PR description but:

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [ ] Longer descriptions should be in documentation or in design decisions.
- [ ] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests
- [ ] I ran all tests locally and everything went fine
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see doc/CODING.md)
- [ ] meta data is updated (e.g. README.md of plugins and doc/METADATA.md)

## Review

Remove the line below and add the "work in progress" label if you do
not want the PR to be reviewed

## Merging

Please add the "ready to merge" label when the build server and you say
that everything is ready to be merged.

### CODE REVIEW

File: src/plugins/yajl/yajl_parse.c [hunk 1]
```diff
@@ -269,6 +269,31 @@ static void elektraYajlParseSuppressEmpty (KeySet * returned, Key * parentKey)
 		}
 		keyDel (lookupKey);
 	}
+
+	ksRewind (returned);
+	Key * cur;
+	while ((cur = ksNext (returned)) != NULL)
+	{
+		// Not sure atm if checking for dirdata is needed or the responsibility of yajl
```
Comment (markus2330) [inline src/plugins/yajl/yajl_parse.c]: @sanssecours can you help here?

Comment (sanssecours) [reply src/plugins/yajl/yajl_parse.c]: Please **ignore the Directory Value plugin**. If it makes your life easier Philipp, then just remove `directoryvalue` from the `infos/needs` line in the plugin contract and re-add it after you fixed issue #2132.

**Comment by PhilippGackstatter** [comment]
So with the recent changes, all non-leaf keys are removed. Empty maps and arrays work as intended.
```sh
kdb mount config.json user/tests/yajl yajl
printf '{"l1": {"l2": {"l3": 1, "l31": {}, "l32": []}}}' > $(kdb file user/tests/yajl)
kdb ls user/tests/yajl
#> user/tests/yajl/l1/l2/l3
#> user/tests/yajl/l1/l2/l31/___empty_map
#> user/tests/yajl/l1/l2/l32
#> user/tests/yajl/l1/l2/l32/###empty_array
```

For arrays I had to add an exception, so that the arrays name does not get removed. Previously, the code removed all non-leaf keys, but the array name is also a non-leaf key. Since `yamlcpp` does not remove the array name, I assume this is intended. Which also makes sense to me, so that applications have an "entry point" for the array.

**Comment by markus2330** [comment]
Thank you! You are right about leaving the array keys. The array syntax, however, changed a bit since yajl was developed, see doc/decisions/array.md

**Pull request merged**